File: python/src/circle_ai/inference/inference.py
"""Inference-layer contracts for CircleAI Python.

Mirrors CircleAI.Inference in the C# port:

    * IChatGenerator   — text generator contract + generate_response_async
                          default-equivalent.
    * IChatGenerator   gains ChatResponse-returning helper.
    * ChatCapability   — flag enum used by IModelSelector.
    * IModelSelector / ModelSelection — capability + device-fit picker.
"""
from __future__ import annotations

import time
from dataclasses import dataclass
from enum import IntFlag
from typing import (
    TYPE_CHECKING,
    AsyncGenerator,
    Optional,
    Protocol,
    Sequence,
    runtime_checkable,
)

from enum import IntEnum

from ..models.models import (
    ChatFragment,
    ChatFragmentKind,
    ChatMessage,
    ChatResponse,
    FinishReason,
)


# ── PowerBudget (RT-11) ───────────────────────────────────────────────────


class PowerBudget(IntEnum):
    """Per-call power budget — how much device energy this generation is worth.

    Mirrors CircleAI.Inference.PowerBudget. The runtime translates the budget
    into a per-call max-tokens cap (and, when fallback chains exist, into a
    model-size pick).
    """

    NONE = 0
    """Opt out of automatic budget control — honour max_tokens literally."""

    LOW = 1
    """Battery-conscious. ~64 token cap, prefers TQ4 KV, smaller model in chain."""

    NORMAL = 2
    """Default balanced behaviour. ~512 token cap. Auto-downgrades to LOW below 15% battery."""

    HIGH = 3
    """Quality-first. ~2048 token cap, full FP16 KV. Auto-throttles on thermal warnings."""

if TYPE_CHECKING:  # pragma: no cover
    from ..device.device_probe import DeviceProbe, DeviceTier


# ── GenerationOptions ────────────────────────────────────────────────────


@dataclass
class GenerationOptions:
    """Knobs for a single generation call."""

    max_tokens: int = 512
    temperature: float = 0.7
    top_p: float = 0.9
    top_k: int = 40
    seed: Optional[int] = None
    stop_sequences: Optional[list[str]] = None

    include_reasoning: bool = True
    """Whether to surface the model's reasoning trace (Qwen3
    ``<think>…</think>``) on the call.

    When ``True`` (default) the generator separates reasoning from the final
    answer: ``ChatResponse.reasoning_content`` gets the reasoning,
    ``ChatResponse.text`` gets the answer. Streaming callers see fragments
    tagged with ``ChatFragmentKind.REASONING``.

    When ``False`` the generator still RUNS reasoning (this is per-call output
    gating, NOT a thinking disable) but the reasoning text is dropped — only
    the final answer reaches the caller. Use this for JSON-strict consumers.
    """

    budget: PowerBudget = PowerBudget.NORMAL
    """(RT-11) Declarative power budget. The runtime maps it to a max-tokens
    cap and (eventually) model size. NORMAL auto-downgrades to LOW when battery
    is below 15%. Use NONE to honour max_tokens literally.
    """

    use_prefix_cache: bool = False
    """(RT-06) Whether the runtime should consult the cross-session prefix
    cache for a warm (model_id, system_prompt) snapshot before resetting the
    model handle. First call populates; subsequent calls reload it instead of
    running the system-prompt prefill.
    """


# ── IChatGenerator ───────────────────────────────────────────────────────


@runtime_checkable
class IChatGenerator(Protocol):
    """Contract for an on-device chat-style text generator."""

    async def generate_async(
        self,
        messages: list[ChatMessage],
        options: Optional[GenerationOptions] = None,
    ) -> str:
        """Generate a complete assistant reply for the given conversation."""
        ...

    async def stream_async(
        self,
        messages: list[ChatMessage],
        options: Optional[GenerationOptions] = None,
    ) -> AsyncGenerator[str, None]:
        """Stream the assistant reply chunk-by-chunk. Content only — any
        reasoning emitted inside ``<think>…</think>`` is filtered out. Use
        :func:`stream_fragments_async` when you also need the reasoning stream.
        """
        ...

    async def save_session_async(self, path: str) -> bool:
        """(RT-02) Save the current model session to ``path``. Returns
        ``True`` on success. Default protocol contract returns ``False`` —
        native generators (MNN-backed) override.
        """
        return False

    async def load_session_async(self, path: str) -> bool:
        """(RT-02) Load a previously-saved session from ``path``. Returns
        ``True`` on success. Default protocol contract returns ``False`` —
        native generators (MNN-backed) override.
        """
        return False


async def generate_response_async(
    generator: IChatGenerator,
    messages: list[ChatMessage],
    options: Optional[GenerationOptions] = None,
) -> ChatResponse:
    """Wrap IChatGenerator.generate_async into a structured ChatResponse.

    Python's Protocol doesn't support default method implementations the way
    C# default-interface-methods do, so this is exposed as a free function.
    Native generators may shadow it with a method that reports exact token
    counts and surfaces ``ChatResponse.reasoning_content`` — this default
    approximates and leaves reasoning_content as ``None``.
    """
    started = time.monotonic()
    text = await generator.generate_async(messages, options)
    latency_ms = (time.monotonic() - started) * 1000.0

    tokens_in = _approx_tokens_messages(messages)
    tokens_out = _approx_tokens(text)

    return ChatResponse(
        text=text,
        tokens_in=tokens_in,
        tokens_out=tokens_out,
        latency_ms=latency_ms,
        finish_reason=FinishReason.STOP,
        reasoning_content=None,
    )


async def stream_fragments_async(
    generator: IChatGenerator,
    messages: list[ChatMessage],
    options: Optional[GenerationOptions] = None,
) -> AsyncGenerator[ChatFragment, None]:
    """Wrap IChatGenerator.stream_async into the fragment-tagged stream.

    Default helper: yields each chunk from ``stream_async`` tagged as
    ``ChatFragmentKind.CONTENT``. Generators that surface reasoning must
    expose their own ``stream_fragments_async`` method that interleaves
    ``REASONING`` fragments — this helper does NOT split ``<think>`` tags
    (that requires generator-level token routing).
    """
    async for chunk in generator.stream_async(messages, options):
        yield ChatFragment(kind=ChatFragmentKind.CONTENT, text=chunk)


def _approx_tokens(text: Optional[str]) -> int:
    if not text:
        return 0
    # Crude 4-chars-per-token approximation; matches the C# fallback.
    return max(1, len(text) // 4)


def _approx_tokens_messages(messages: Sequence[ChatMessage]) -> int:
    return sum(_approx_tokens(m.content) for m in messages)


# ── ChatCapability + IModelSelector ──────────────────────────────────────


class ChatCapability(IntFlag):
    """Capabilities a chat model declares.

    Mirrors CircleAI.Inference.ChatCapability. Selectors filter by these
    flags so consumers can request "vision + tools" and only get a model
    that handles both.
    """

    NONE = 0
    DEFAULT = 1
    TOOLS = 2
    VISION = 4
    LONG_CONTEXT = 8
    REASONING = 16


@dataclass(frozen=True)
class ModelSelection:
    """One selector result. tier is the device tier the pick was sized for."""

    model_id: str
    requires_download: bool
    estimated_bytes: int
    tier: "DeviceTier"


@runtime_checkable
class IModelSelector(Protocol):
    """Picks a model that fits the device + the requested capabilities."""

    def best_fit(
        self,
        probe: "DeviceProbe",
        required: ChatCapability = ChatCapability.DEFAULT,
    ) -> ModelSelection:
        """Returns the highest-quality entry that satisfies every flag in
        `required` AND has MinRamGb <= probe RAM AND MinStorageGb <= free."""
        ...

    def all_candidates(self, probe: "DeviceProbe") -> list[ModelSelection]:
        """Every selection candidate in registry order — diagnostics use."""
        ...

File: kotlin/src/main/kotlin/com/bhengubv/circleai/inference/Inference.kt
// Inference.kt
//
// Kotlin port of Circle.AI.Inference portable layer.
//
// Covers:
//   GenerationOptions  — knobs for a single generation call
//   IChatGenerator     — contract for an on-device chat-style text generator
//
// Note: ChatMessage is declared in com.bhengubv.circleai.models.Models to avoid
// duplication. The inference layer imports it from there.

package com.bhengubv.circleai.inference

import com.bhengubv.circleai.models.ChatFragment
import com.bhengubv.circleai.models.ChatFragmentKind
import com.bhengubv.circleai.models.ChatMessage
import com.bhengubv.circleai.models.ChatResponse
import com.bhengubv.circleai.models.FinishReason
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.io.File
import java.time.Instant
import kotlin.math.max

// ---------------------------------------------------------------------------
// GenerationOptions
// ---------------------------------------------------------------------------

/**
 * Knobs for a single generation call.
 * All fields have sensible defaults and are immutable after construction.
 *
 * Mirrors CircleAI.Inference.GenerationOptions (C# is the spec): `MaxTokens`
 * defaults to 512, and the full knob set includes top-k, an optional seed, the
 * declarative [PowerBudget], and the prefix-cache opt-in.
 */
data class GenerationOptions(
    /** Maximum number of new tokens to produce. */
    val maxTokens: Int = 512,
    /** Sampling temperature. 0 = greedy; higher = more random. */
    val temperature: Float = 0.7f,
    /** Nucleus sampling cutoff (top-p). 1.0 disables. */
    val topP: Float = 0.9f,
    /** Top-k cutoff. 0 disables. */
    val topK: Int = 40,
    /** Optional RNG seed. `null` means non-deterministic. */
    val seed: Int? = null,
    /**
     * Optional substrings that will end generation when matched in the emitted
     * output (e.g. role-tag boundaries).
     */
    val stopSequences: List<String> = emptyList(),
    /** Optional system prompt to prepend before the conversation messages. */
    val systemPrompt: String = "",
    /**
     * Whether to surface the model's reasoning trace (Qwen3
     * `<think>…</think>`) on the call.
     *
     * When `true` (default) the generator separates reasoning from the final
     * answer: `ChatResponse.reasoningContent` gets the reasoning,
     * `ChatResponse.text` gets the answer. Streaming callers see fragments
     * tagged with `ChatFragmentKind.REASONING`.
     *
     * When `false` the generator still runs reasoning (this is per-call output
     * gating, NOT a thinking disable) but the reasoning text is dropped — only
     * the final answer reaches the caller. Use this for JSON-strict consumers.
     */
    val includeReasoning: Boolean = true,
    /**
     * (RT-11) Declarative power budget for this call. The runtime maps the
     * budget to context size, KV compression, and decode token limit. Default
     * [PowerBudget.Normal] auto-downgrades to `Low` below 15% battery. Pass
     * [PowerBudget.None] to honour [maxTokens] literally.
     */
    val budget: PowerBudget = PowerBudget.Normal,
    /**
     * (RT-06) Whether the runtime should consult the cross-session prefix cache
     * for a warm (modelId, systemPrompt) snapshot before resetting the model
     * handle. Default `false`.
     */
    val usePrefixCache: Boolean = false,
)

// ---------------------------------------------------------------------------
// IChatGenerator
// ---------------------------------------------------------------------------

/**
 * Contract for an on-device chat-style text generator. Implementations own
 * native model state and must be [AutoCloseable].
 */
interface IChatGenerator : AutoCloseable {
    /**
     * Generates a complete assistant reply for the given conversation.
     */
    suspend fun generateAsync(
        messages: List<ChatMessage>,
        opts: GenerationOptions = GenerationOptions()
    ): String

    /**
     * Streams the assistant reply token-by-token (or piece-by-piece) as it is
     * decoded. Each emitted string is the next chunk to append to the output —
     * callers should concatenate them in order. Content only — any reasoning
     * inside `<think>…</think>` is filtered out. Use [streamFragmentsAsync]
     * when you also need the reasoning stream.
     */
    fun streamAsync(
        messages: List<ChatMessage>,
        opts: GenerationOptions = GenerationOptions()
    ): Flow<String>

    /**
     * Fragment-aware streaming variant. Yields each piece tagged as either
     * [ChatFragmentKind.CONTENT] or [ChatFragmentKind.REASONING] so the caller
     * can route the model's `<think>` block into a separate `reasoning_content`
     * field (o1 / DeepSeek style).
     *
     * Default implementation wraps [streamAsync] and tags every chunk as
     * `CONTENT`; generators that surface reasoning override this method.
     */
    fun streamFragmentsAsync(
        messages: List<ChatMessage>,
        opts: GenerationOptions = GenerationOptions()
    ): Flow<ChatFragment> =
        streamAsync(messages, opts).map { ChatFragment(ChatFragmentKind.CONTENT, it) }

    /**
     * (RT-02) Save the current model session — KV cache + history — to [path]
     * so the conversation can survive an OOM kill and resume later via
     * [loadSessionAsync].
     *
     * Default implementation writes a portable marker file containing the
     * generator type name + a UTC timestamp so callers always get a
     * non-throwing round-trip. Native generators (Qwen, KimiVl) override to
     * call the MNN session primitives. Returns `true` on success.
     */
    suspend fun saveSessionAsync(path: String): Boolean {
        require(path.isNotBlank()) { "path required" }
        val marker = "circleai-session-marker\ntype:${this::class.qualifiedName}\nsaved_utc:${Instant.now()}\n"
        File(path).writeText(marker)
        return true
    }

    /**
     * (RT-02) Load a previously-saved session from [path]. Default
     * implementation verifies the marker file written by the default
     * [saveSessionAsync]. Native generators override to restore real KV-cache
     * state. Returns `true` on success.
     */
    suspend fun loadSessionAsync(path: String): Boolean {
        require(path.isNotBlank()) { "path required" }
        val f = File(path)
        if (!f.exists()) return false
        return f.readText().startsWith("circleai-session-marker")
    }

    /**
     * Structured-response variant: returns the assistant reply alongside token
     * counts, finish reason, and latency. Default implementation wraps
     * [generateAsync] with an approximate token count (word split) and
     * [FinishReason.STOP]; native generators override to report exact
     * native-reported values (and to surface [ChatResponse.reasoningContent]).
     */
    suspend fun generateResponseAsync(
        messages: List<ChatMessage>,
        opts: GenerationOptions = GenerationOptions(),
    ): ChatResponse {
        val started = System.nanoTime()
        val text = generateAsync(messages, opts)
        val latencyMs = (System.nanoTime() - started) / 1_000_000.0
        val tokensIn = messages.sumOf { approximateTokenCount(it.content) }
        val tokensOut = approximateTokenCount(text)
        return ChatResponse(
            text = text,
            tokensIn = tokensIn,
            tokensOut = tokensOut,
            latencyMs = latencyMs,
            finishReason = FinishReason.STOP,
        )
    }

    private fun approximateTokenCount(text: String?): Int {
        if (text.isNullOrEmpty()) return 0
        // Crude approximation — 1 token ~= 4 chars in English.
        return max(1, text.length / 4)
    }
}
